TailLight/device.cpp

```diff
@@ -1,52 +1,77 @@
+    //KdPrint(("TailLight: PnpNotifyDeviceInterfaceChange enter\n"));
+    NT_ASSERTMSG("WDFDEVICE not passed in!", pvContext);
 
-    NTSTATUS status = SetFeatureColor(device, 0);
-    if (!NT_SUCCESS(status)) {
-        KdPrint(("TailLight: EvtSetBlackTimer failure NTSTATUS=0x%x\n", status));
-        return;
+    if (pvNotificationStructure == NULL) {
+        return STATUS_SUCCESS;
+    }
+
+    PDEVICE_INTERFACE_CHANGE_NOTIFICATION pDevInterface =
+        (PDEVICE_INTERFACE_CHANGE_NOTIFICATION)pvNotificationStructure;
+
+    ASSERT(IsEqualGUID(*(_GUID*)&(pDevInterface->InterfaceClassGuid),
+        *(_GUID*)&GUID_DEVINTERFACE_HID));
+
+    {     
+        // Assumption: Device will arrive before removal.
+        if (IsEqualGUID(*(LPGUID)&(pDevInterface->Event),
+            *(LPGUID)&GUID_DEVICE_INTERFACE_ARRIVAL)) {
+
+            // Opening a device may trigger PnP operations. Ensure that either a
+            // timer or a work item is used when opening up a device.
+            // Refer to p356 of Oney and IoGetDeviceObjectPointer.
+            //
+            // NOTE: It is possible for us to get blocked waiting for a system
+            // thread. One solution would be to use a timer that spawns a
+            // system thread on timeout vs. a wait loop.
+
+
+            // Always must return success, so ignore result.
+            CreateWorkItemForIoTargetOpenDevice((WDFDEVICE)pvContext,
+                *pDevInterface->SymbolicLinkName);
+        }
     }
 
-    KdPrint(("TailLight: EvtSetBlackTimer end\n"));
+    return STATUS_SUCCESS;
 }
 
+
 NTSTATUS EvtSelfManagedIoInit(WDFDEVICE device) {
-    // Initialize tail-light to black to have control over HW state
-    WDF_TIMER_CONFIG timerCfg = {};
-    WDF_TIMER_CONFIG_INIT(&timerCfg, EvtSetBlackTimer);
 
-    WDF_OBJECT_ATTRIBUTES attribs = {};
-    WDF_OBJECT_ATTRIBUTES_INIT(&attribs);
-    attribs.ParentObject = device;
-    attribs.ExecutionLevel = WdfExecutionLevelPassive; // required to access HID functions
+    WDFDRIVER driver = WdfDeviceGetDriver(device);
+    DRIVER_CONTEXT* driverContext = WdfObjectGet_DRIVER_CONTEXT(driver);
 
-    WDFTIMER timer = nullptr;
-    NTSTATUS status = WdfTimerCreate(&timerCfg, &attribs, &timer);
-    if (!NT_SUCCESS(status)) {
-        KdPrint(("WdfTimerCreate failed 0x%x\n", status));
-        return status;
-    }
+    //TRACE_FN_ENTRY;
 
-    status = WdfTimerStart(timer, 0); // no wait
-    if (!NT_SUCCESS(status)) {
-        KdPrint(("WdfTimerStart failed 0x%x\n", status));
-        return status;
-    }
+    NTSTATUS status = STATUS_SUCCESS;
+
+    status = IoRegisterPlugPlayNotification(
+        EventCategoryDeviceInterfaceChange,
+        PNPNOTIFY_DEVICE_INTERFACE_INCLUDE_EXISTING_INTERFACES,
+        (PVOID)&GUID_DEVINTERFACE_HID,
+        WdfDriverWdmGetDriverObject(driver),
+        PnpNotifyDeviceInterfaceChange,
+        (PVOID)device,
+        &driverContext->pnpDevInterfaceChangedHandle
```

Comment:
There seem to be a mismatch with the way `pnpDevInterfaceChangedHandle` is registered & unregistered. It's currently registered when attaching a new _device_ and unregistered when unloading the _driver_. This could cause `pnpDevInterfaceChangedHandle` to be registered twice if inserting two identical IntelliMouse's and only unregistered once after both are disconnected.

Wouldn't it be better to either:
* Move the unregistration code to when the _device_ is removed/disconnected, or
* Move the registration code to when the _driver_ is loaded.

Reply:
Thanks for improving on this.